infrastructure/lib/mail.ts

```diff
@@ -0,0 +1,55 @@
+import * as cdk from '@aws-cdk/core';
+import * as route53 from '@aws-cdk/aws-route53';
+import { DomainName, MXSPFRecords, MXReconds } from './variables';
+import { SSMParameterReader } from './ssm-parameter-reader';
+import { HostedZoneIdParam, HostedZoneNameParam } from './domain';
+
+export class MailStack extends cdk.Stack {
+	constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
+		super(scope, id, props);
+
+		const hostedZoneIdReader = new SSMParameterReader(
+			this,
+			'Route53HostedZoneIdReader',
+			{
+				parameterName: HostedZoneIdParam,
+				region: 'us-east-1',
+			}
+		);
```

Comment:
There's actually a fair bit of code duplication between these classes.

Let's say for example we need to change the region all of our stack is using, `us-east-1`. How many places would we need to make this change? I believe I count 6. What happens if one is missed?

It wouldn't be terribly difficult to take this one level of abstraction deeper and create a factory wrapper of sorts for example with all the `hostedZone` code. We can do the same with lots more of this code too.

I don't even know how to write gross class-oriented JS so I'm just going to do it with functional syntax lol

```js
// infrastructure/lib/factories/hostedZone

const hostedZoneFactory = (stackInstance, ...) => { // stackInstance is the 'this' keyword passed in
  const hostedZoneIdReader = ...
  const hostedZoneNameReader = ....

  return {
    hostedZoneIdReader,
    hostedZoneNameReader,
    ...
  }
}
```

```js
// usage

import hostedZoneFactory from 'infrastructure/lib/factories/hostedZone';

export class MailStack extends cdk.Stack {
  constructor(...) {
    ...
    const hostedZone = hostedZoneFactory(this, ...);

    // use hostedZone.hostedZoneIdReader wherever needed, etc.
    
   // alternatively
    const { hostedZoneIdReader, hostedZoneNameReader, ... } = hostedZoneFactory(this, ...);
```

Now we have a hostedzone "class" that we can use throughout these classes at a higher level of abstraction and we only need to make updates to it in one place.

On another note, things like `'us-east-1'` should be placed in a constants file so again there is only ONE place for them to be modified.

```js
region: regionConstants.REGION,
```

Reply:
To the point of having different levels of abstraction, adding factories like this introduces a new level, where the `frontendStack` class would be interacting with these factories now, which would in turn be interacting with the CDK stack. With this example the idea becomes clear I hope, we have
```
// before
FrontendStack class
 - HostedZone from CDK
 - CloudFront from CDK
 - AnythingElse from CDK

// after
FrontendStack class
  - HostedZoneFactory
      - HostedZone from CDK
  - CloudFrontFactory
      - CloudFront from CDK
  - AnythingElseFactory
      -AnythingElse from CDK
```

We hide the implementation details in the factories, or classes, whatever you want them to be. This removes responsibilities from the FrontendStack class. You can say the factories are operating at a lower level of abstraction and the FrontendStack class is operating at a higher level of abstraction. The more levels we introduce the more we separate concerns and create reusable code, though this can be obviously taken too far.

Hope this is clear!